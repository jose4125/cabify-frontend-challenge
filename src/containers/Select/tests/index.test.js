import React from 'react';
import { shallow, mount } from 'enzyme';

import { Select } from '../index';

describe('Select Field', () => {
  const mockSetCountrySelected = jest.fn();

  const props = {
    isOpen: false,
    codes: [],
    countryCodeSelected: '+93',
    setCountrySelected: mockSetCountrySelected,
    counter: 0,
    optionSelectedElement: null
  };

  const wrapper = shallow(<Select {...props} />);

  it('should render with out crashing', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should have a chevron', () => {
    expect(wrapper.find('.select__chevron').exists()).toBe(true);
  });

  it('should have a country code text', () => {
    expect(wrapper.find('.select__button').text()).toBe(
      props.countryCodeSelected
    );
  });

  it('should have the list container', () => {
    expect(wrapper.find('ul.dropdown__wrapper').exists()).toBe(true);
  });

  it('should not have an isOpen and focus classes', () => {
    expect(wrapper.prop('className')).not.toContain('isOpen focus');
  });
});

describe('Select Field mounted', () => {
  const mockSetCountrySelected = jest.fn();
  const mockHandleOpenCloseDropdown = jest.fn();
  const mockHandleClickOutside = jest.fn();
  let props;
  let wrapper;

  beforeEach(() => {
    props = {
      isOpen: true,
      codes: [
        {
          name: 'Afghanistan',
          dial_code: '+93',
          code: 'AF'
        },
        {
          name: 'Albania',
          dial_code: '+355',
          code: 'AL'
        },
        {
          name: 'Algeria',
          dial_code: '+213',
          code: 'DZ'
        }
      ],
      countryCodeSelected: '+93',
      setCountrySelected: mockSetCountrySelected,
      handleOpenCloseDropdown: mockHandleOpenCloseDropdown,
      handleClickOutside: mockHandleClickOutside,
      counter: 0,

      optionSelectedElement: null
    };

    wrapper = shallow(<Select {...props} />);
  });

  it('should call setCountrySelected when was mounted', () => {
    expect(mockSetCountrySelected).toHaveBeenCalledWith(
      props.countryCodeSelected
    );
    expect(mockSetCountrySelected).toHaveBeenCalledTimes(1);
  });

  it('should call handleOpenCloseDropdown when the button was clicked', () => {
    wrapper.find('.select__button').simulate('click');
    expect(mockHandleOpenCloseDropdown).toHaveBeenCalled();
  });

  it('should have an isOpen and focus classes', () => {
    expect(wrapper.prop('className')).toContain('isOpen focus');
  });

  it('should handle the down arrow key event', () => {
    wrapper.find('.select__button').simulate('keydown', { keyCode: 40 });
    expect(mockSetCountrySelected).toHaveBeenCalledWith(
      props.codes[1].dial_code
    );
  });
});
