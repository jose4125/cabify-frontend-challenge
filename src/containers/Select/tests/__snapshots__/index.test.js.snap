// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select Field should render with out crashing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Select
    codes={Array []}
    counter={0}
    countryCodeSelected="+93"
    isOpen={false}
    optionSelectedElement={null}
    setCountrySelected={
      [MockFunction] {
        "calls": Array [
          Array [
            "+93",
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="input select"
        >
          <p
            className="select__label"
          >
            Prefix
          </p>
          <button
            className="select__button"
            onKeyDown={[Function]}
            type="button"
          >
            +93
            <img
              alt="chevron"
              className="select__chevron"
              src="arrow.svg"
            />
          </button>
        </div>,
        <div
          className="dropdown"
        >
          <ul
            className="dropdown__wrapper"
          />
        </div>,
      ],
      "className": "formField-input formField-input--select  active  col col12",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <p
              className="select__label"
            >
              Prefix
            </p>,
            <button
              className="select__button"
              onKeyDown={[Function]}
              type="button"
            >
              +93
              <img
                alt="chevron"
                className="select__chevron"
                src="arrow.svg"
              />
            </button>,
          ],
          "className": "input select",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Prefix",
              "className": "select__label",
            },
            "ref": null,
            "rendered": "Prefix",
            "type": "p",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                "+93",
                <img
                  alt="chevron"
                  className="select__chevron"
                  src="arrow.svg"
                />,
              ],
              "className": "select__button",
              "onClick": undefined,
              "onKeyDown": [Function],
              "type": "button",
            },
            "ref": null,
            "rendered": Array [
              "+93",
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "alt": "chevron",
                  "className": "select__chevron",
                  "src": "arrow.svg",
                },
                "ref": null,
                "rendered": null,
                "type": "img",
              },
            ],
            "type": "button",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <ul
            className="dropdown__wrapper"
          />,
          "className": "dropdown",
        },
        "ref": Object {
          "current": null,
        },
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": undefined,
            "className": "dropdown__wrapper",
          },
          "ref": Object {
            "current": null,
          },
          "rendered": null,
          "type": "ul",
        },
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="input select"
          >
            <p
              className="select__label"
            >
              Prefix
            </p>
            <button
              className="select__button"
              onKeyDown={[Function]}
              type="button"
            >
              +93
              <img
                alt="chevron"
                className="select__chevron"
                src="arrow.svg"
              />
            </button>
          </div>,
          <div
            className="dropdown"
          >
            <ul
              className="dropdown__wrapper"
            />
          </div>,
        ],
        "className": "formField-input formField-input--select  active  col col12",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <p
                className="select__label"
              >
                Prefix
              </p>,
              <button
                className="select__button"
                onKeyDown={[Function]}
                type="button"
              >
                +93
                <img
                  alt="chevron"
                  className="select__chevron"
                  src="arrow.svg"
                />
              </button>,
            ],
            "className": "input select",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Prefix",
                "className": "select__label",
              },
              "ref": null,
              "rendered": "Prefix",
              "type": "p",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "+93",
                  <img
                    alt="chevron"
                    className="select__chevron"
                    src="arrow.svg"
                  />,
                ],
                "className": "select__button",
                "onClick": undefined,
                "onKeyDown": [Function],
                "type": "button",
              },
              "ref": null,
              "rendered": Array [
                "+93",
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "alt": "chevron",
                    "className": "select__chevron",
                    "src": "arrow.svg",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "img",
                },
              ],
              "type": "button",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <ul
              className="dropdown__wrapper"
            />,
            "className": "dropdown",
          },
          "ref": Object {
            "current": null,
          },
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": undefined,
              "className": "dropdown__wrapper",
            },
            "ref": Object {
              "current": null,
            },
            "rendered": null,
            "type": "ul",
          },
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": false,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
